engines/prover.cpp

```diff
@@ -68,10 +70,284 @@ void Prover::initialize()
+  Term cur;
+
+  while (!open_terms.empty()) {
+    cur = open_terms.back();
+    open_terms.pop_back ();
+
+    if (visited_terms.find(cur) == visited_terms.end()) {
+      // cache 'cur' and push its children
+      visited_terms.insert(cur);
+
+      cout << "  visiting term: " << cur << "\n";
+      if (cur->is_symbol())
+        cout << "    ..is symbol\n";
+      
+      for (auto child : cur) {
+        cout << "    pushing child: " << child << "\n";
+        open_terms.push_back(child);
+      }
+    }
+  }
+}
+
+/* For debugging only. */
+void Prover::print_coi_info()
+{
+  cout << "TEST PRINT COI\n";
+  cout << "bad_ term: " << bad_ << "\n";
+  print_term_dfs(bad_);
+
+  cout << "init_ term: " << ts_.init() << "\n";
+  print_term_dfs(ts_.init());
+
+  cout << "trans_ term: " << ts_.trans() << "\n";
+  print_term_dfs(ts_.trans());
+
+  cout << "input vars: \n";
+  for (auto inputvar : ts_.inputvars())
+    cout << "  " << inputvar << "\n";
+
+  cout << "state vars: \n";
+  for (auto statevar : ts_.statevars())
+    cout << "  " << statevar << "\n";
+
+  cout << "constraints: \n";
+  for (auto constr : ts_.constraints())
+    cout << "  " << constr << "\n";  
+}
+
+/* Add 'term' to 'set' if it does not already appear there. */
+void Prover::collect_coi_term(UnorderedTermSet & set, const Term & term)
+{
+  if (set.find(term) == set.end())
+    set.insert(term);
+}
+
+/* Traverse 'term', collect state/input variables, and add them to
+   global sets 'new_coi_state_vars' and 'new_coi_input_vars'. */
+void Prover::compute_term_coi(const Term & term,
+                              UnorderedTermSet & new_coi_state_vars,
+                              UnorderedTermSet & new_coi_input_vars)
+{
+  assert(term != NULL);
+  TermVec open_terms;
+  open_terms.push_back (term);
+  Term cur;
+
+  /* Depth-first search of term structure 'term'. */
+  while (!open_terms.empty()) {
+    cur = open_terms.back();
+    open_terms.pop_back ();
+
+    /* Use global set 'coi_visited_terms' here to avoid visiting terms
+       multiple times when we call this function on different terms. */
+    if (coi_visited_terms_.find(cur) == coi_visited_terms_.end()) {
+      /* Cache 'cur' and push its children. */
+      coi_visited_terms_.insert(cur);
+      logger.log(3, "  visiting COI term: {}", cur);
+      if (cur->is_symbol()) {
+        logger.log(3, "    ..is symbol");
+        if (ts_.statevars().find(cur) != ts_.statevars().end()) {
+          assert(ts_.inputvars().find(cur) == ts_.inputvars().end());
+          assert(!ts_.is_next_var(cur));
+          logger.log(3, "collect COI statevar {}", cur);
+          collect_coi_term(new_coi_state_vars, cur);
+        }
+        else if (ts_.inputvars().find(cur) != ts_.inputvars().end()) {
+          assert(!ts_.is_curr_var(cur));
+          assert(!ts_.is_next_var(cur));
+          logger.log(3, "collect COI inputvar {}", cur);
+          collect_coi_term(new_coi_input_vars, cur);
+        }
```

Comment:
What about the next state variable case? Constraints can have next state variables because if it's an invariant, the constraint is added both over current vars and next vars. Is it safe to ignore those cases because there's always a current state version of the same invariant?

Reply:
>  Is it safe to ignore those cases because there's always a current state version of the same invariant?
Yes, that is exactly what I was thinking. That should work since we collect **all** state variables in **all** constraints unconditionally, i.e., without checking whether they are syntactically related to the bad-state term.